internal/backend/grpc/serviceaccount.go

```diff
@@ -10,22 +10,317 @@ import (
+	}
+
+	key, err := validatePGPPublicKey(
+		[]byte(req.GetArmoredPgpPublicKey()),
+		pgp.WithMaxAllowedLifetime(auth.ServiceAccountMaxAllowedLifetime),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	publicKeyResource := authres.NewPublicKey(resources.DefaultNamespace, key.id)
+	publicKeyResource.Metadata().Labels().Set(authres.LabelPublicKeyUserID, identity.TypedSpec().Value.UserId)
+
+	publicKeyResource.TypedSpec().Value.PublicKey = key.data
+	publicKeyResource.TypedSpec().Value.Expiration = timestamppb.New(key.expiration)
+	publicKeyResource.TypedSpec().Value.Role = user.TypedSpec().Value.GetRole()
+
+	publicKeyResource.TypedSpec().Value.Confirmed = true
+
+	publicKeyResource.TypedSpec().Value.Identity = &specs.Identity{
+		Email: id,
+	}
+
+	err = s.omniState.Create(ctx, publicKeyResource)
+	if err != nil {
+		return nil, err
+	}
+
+	return &management.RenewServiceAccountResponse{PublicKeyId: key.id}, nil
+}
+
+func (s *managementServer) ListServiceAccounts(ctx context.Context, _ *emptypb.Empty) (*management.ListServiceAccountsResponse, error) {
+	_, err := s.authCheckGRPC(ctx, auth.WithRole(role.Admin))
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = actor.MarkContextAsInternalActor(ctx)
+
+	identityList, err := safe.StateListAll[*authres.Identity](
+		ctx,
+		s.omniState,
+		state.WithLabelQuery(resource.LabelExists(authres.LabelIdentityTypeServiceAccount)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccounts := make([]*management.ListServiceAccountsResponse_ServiceAccount, 0, identityList.Len())
+
+	for iter := identityList.Iterator(); iter.Next(); {
+		identity := iter.Value()
+
+		user, err := safe.StateGet[*authres.User](ctx, s.omniState, authres.NewUser(resources.DefaultNamespace, identity.TypedSpec().Value.UserId).Metadata())
+		if err != nil {
+			return nil, err
+		}
+
+		publicKeyList, err := safe.StateListAll[*authres.PublicKey](
+			ctx,
+			s.omniState,
+			state.WithLabelQuery(resource.LabelEqual(authres.LabelPublicKeyUserID, user.Metadata().ID())),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		publicKeys := make([]*management.ListServiceAccountsResponse_ServiceAccount_PgpPublicKey, 0, publicKeyList.Len())
+
+		for keyIter := publicKeyList.Iterator(); keyIter.Next(); {
+			key := keyIter.Value()
+
+			publicKeys = append(publicKeys, &management.ListServiceAccountsResponse_ServiceAccount_PgpPublicKey{
+				Id:         key.Metadata().ID(),
+				Armored:    string(key.TypedSpec().Value.GetPublicKey()),
+				Expiration: key.TypedSpec().Value.GetExpiration(),
+			})
+		}
+
+		sa, isSa := pkgaccess.ParseServiceAccountFromFullID(identity.Metadata().ID())
+		if !isSa {
+			return nil, fmt.Errorf("unexpected service account ID %q", identity.Metadata().ID())
+		}
+
+		name := sa.NameWithPrefix()
+
+		serviceAccounts = append(serviceAccounts, &management.ListServiceAccountsResponse_ServiceAccount{
+			Name:          name,
+			PgpPublicKeys: publicKeys,
+			Role:          user.TypedSpec().Value.GetRole(),
```

Comment:
maybe lets add IsCloudProvider flag to this response

Reply:
I can, but I think it would actually be redundant to this `Role` field, as only cloud providers can have `CloudProvider` role (it is enforced by validations). So the client can simply check for that instead, no?

Reply:
ah, right.